<a href="https://colab.research.google.com/github/nrmlptr/Analisis-Data-Dengan-Python/blob/main/Latihan_Exploratory_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Tujuan dari latihan ini
1. mengumpulkan seluruh data yang dibutuhkan;
2. menilai kualitas dari data yang telah dikumpulkan; dan
3. membersihkan data tersebut sehingga siap untuk dianalisis.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

 selanjutnya pada tahap ini, kita akan mengumpulkan semua data yang dibutuhkan.

In [2]:
# load tabel customer
customers_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/customers.csv")
customers_df.head()

customer_id customer_name             gender  age  \
0            1       fulan 1             Female   30   
1            2       fulan 2  Prefer not to say   69   
2            3       fulan 3  Prefer not to say   59   
3            4       fulan 4  Prefer not to say   67   
4            5       fulan 5  Prefer not to say   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country  
0            Northern Territory  Australia  
1               South Australia  Australia  
2  Australian Capital Territory  Australia  
3                    Queensland  Australia  
4                    Queensland  Australia

In [3]:
# load tabel order
orders_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/orders.csv")
orders_df.head()

order_id  customer_id  payment order_date delivery_date
0         1           64    30811  2021-8-30    2021-09-24
1         2          473    50490   2021-2-3    2021-02-13
2         3          774    46763  2021-10-8    2021-11-03
3         4          433    39782   2021-5-6    2021-05-19
4         5          441    14719  2021-3-23    2021-03-24

In [4]:
# load tabel product
product_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/products.csv")
product_df.head()

product_id product_type  product_name size colour  price  quantity  \
0           0        Shirt  Oxford Cloth   XS    red    114        66   
1           1        Shirt  Oxford Cloth    S    red    114        53   
2           2        Shirt  Oxford Cloth    M    red    114        54   
3           3        Shirt  Oxford Cloth    L    red    114        69   
4           4        Shirt  Oxford Cloth   XL    red    114        47   

                                    description  
0  A red coloured, XS sized, Oxford Cloth Shirt  
1   A red coloured, S sized, Oxford Cloth Shirt  
2   A red coloured, M sized, Oxford Cloth Shirt  
3   A red coloured, L sized, Oxford Cloth Shirt  
4  A red coloured, XL sized, Oxford Cloth Shirt

In [5]:
# load tabel sales
sales_df = pd.read_csv("https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/main/DicodingCollection/sales.csv")
sales_df.head()

sales_id  order_id  product_id  price_per_unit  quantity  total_price
0         0         1         218             106         2        212.0
1         1         1         481             118         1        118.0
2         2         1           2              96         3        288.0
3         3         1        1002             106         2        212.0
4         4         1         691             113         3        339.0

setelah berhasil memuat semua data yang dibutuhkan. Tahap selanjutnya ialah menilai kualitas dari data tersebut.
tahap kedua dalam proses data wrangling yaitu assessing data. pada proses ini kualitas dari keseluruhan data yang digunakan akan dinilai. Penilaian ini bertujuan untuk melihat berbagai permasalahan yang ada dalam data tersebut.



In [6]:
# nilai data customer_df
# periksa tipe data data dari tiap kolom yang terdapat dalam customers_df. Proses ini dapat dilakukan menggunakan method info()
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


terdapat perbedaan nilai pada kolom gender

In [7]:
# cari tau missing value pada kolom gender
customers_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
dtype: int64

In [8]:
# periksa duplikasi pada data
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  6


selanjutnya memeriksa parameter statistik dari kolom numerik yang terdapat dalam customers_df. Untuk mempermudah pemeriksaan, gunakan method describe(). Method tersebut akan menampilkan ringkasan parameter statistik (mean, median, dll.) dari kolom numerik pada sebuah DataFrame.

In [9]:
customers_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    50.929494  5012.538232
std     288.673238    30.516299  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    65.000000  7493.500000
max    1000.000000   700.000000  9998.000000

In [10]:
# nilai data order_df
# periksa tipe data dari tiap kolom menggunakan method info().
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


terdapat kesalahan dalam tipe data order_date dan delivery_date karena seharusnya bukan object tapi datetime

In [11]:
# periksa duplikasi dan ringkasan parameter statistik dari kolom numerik pada orders_df.
print("Jumlah duplikasi: ",orders_df.duplicated().sum())
orders_df.describe()

Jumlah duplikasi:  0


order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000

In [12]:
# nilai data product_df
# periksa tipe data tiap kolom dalam product_df telah sesuai.
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.2+ KB


In [13]:
# periksa duplikasi dan ringkasan parameter statistik dari product_df.
print("Jumlah duplikasi: ", product_df.duplicated().sum())
product_df.describe()

Jumlah duplikasi:  6


product_id        price     quantity
count  1266.000000  1266.000000  1266.000000
mean    627.926540   105.812006    60.138231
std     363.971586     9.715611    11.682791
min       0.000000    90.000000    40.000000
25%     313.250000    95.250000    50.000000
50%     626.500000   109.000000    60.000000
75%     942.750000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

terdapat duplikasi pada produt_df sejumlah 6.

In [14]:
# nilai data sales_df
# periksa tipe data dan jumlah data tiap kolom dalam sales_df
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


dalam tabel sales_df terdapat missing value pada kolom total_price, mari periksa

In [15]:
sales_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

terdapat missing value sebanyak 10 pada kolom total_price

In [16]:
# periksa duplikasi dan ringkasan parameter statistik dari sales_df.
print("Jumlah duplikasi: ", sales_df.duplicated().sum())
sales_df.describe()

Jumlah duplikasi:  0


sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

 selanjutnya kita ada dalam proses terakhir dalam data wrangling yaitu pembersihan atau cleaning data. Pada tahap ini, kita akan membersihkan berbagai masalah yang telah teridentifikasi dalam proses assessing data.

Membersihkan Data customers_df
Berdasarkan hasil proses assessing data, diketahui bahwa terdapat tiga masalah yang dijumpai dalam customer_df, yaitu duplicate data, missing value, dan inaccurate value.

In [17]:
# Menghilangkan duplicate data
customers_df.drop_duplicates(inplace=True)

In [18]:
# periksa kembali duplikasi data
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


In [19]:
# Menangani missing value
# gunakan teknik filtering untuk mengetahui data yang mengandung missing value
customers_df[customers_df.gender.isna()]

customer_id customer_name gender  age                     home_address  \
38            39      fulan 39    NaN   80  7440 Cameron Estate DrSuite 628   
167          168     fulan 168    NaN   27         2781 Berge MallSuite 452   
322          322     fulan 322    NaN   30        593 Becker CircleApt. 333   
393          393     fulan 393    NaN   34          5158 Levi HillSuite 531   
442          442     fulan 442    NaN   26           5157 Feil RoadApt. 633   
722          720     fulan 720    NaN   40         31 Jordan ParadeApt. 400   
745          743     fulan 743    NaN   57   09 Christopher StreetSuite 967   
773          771     fulan 771    NaN   74     7367 Wright JunctionApt. 773   
798          795     fulan 795    NaN   49          487 Summer MewsApt. 874   
801          798     fulan 798    NaN   56           27 Aiden KnollApt. 875   
825          822     fulan 822    NaN   59        41 Jenkins KnollSuite 438   
859          855     fulan 855    NaN   55       603 O'keefe KnollSuite 782   
863          859     fulan 859    NaN   38            32 Isla GroveApt. 078   
914          909     fulan 909    NaN   62    976 Murray Station StApt. 036   
934          929     fulan 929    NaN   68           394 Lily HillSuite 153   
948          943     fulan 943    NaN   64       3117 Heller PlaceSuite 149   
952          946     fulan 946    NaN   24      8227 Nicholas HillSuite 150   
994          988     fulan 988    NaN   35   1130 Turner Estate DrSuite 925   

     zip_code                   city                         state    country  
38       4622  North Victoriachester            Northern Territory  Australia  
167      1975         North Leoburgh             Western Australia  Australia  
322      1640             Jacobiview             Western Australia  Australia  
393      1474             Johnsburgh                    Queensland  Australia  
442      7249             Port Chloe               New South Wales  Australia  
722      1380             West Henry               South Australia  Australia  
745      6226         Lake Lukemouth             Western Australia  Australia  
773      8882               Kuhntown                      Victoria  Australia  
798      1712            East Hayden  Australian Capital Territory  Australia  
801      6531               Port Sam  Australian Capital Territory  Australia  
825      2588        Lake Andrewport               South Australia  Australia  
859      8822        Port Dylanmouth                      Tasmania  Australia  
863      7711            Rosechester               New South Wales  Australia  
914      3227               Langfort                      Tasmania  Australia  
934      2353             Beahanfurt            Northern Territory  Australia  
948       822           North Elijah               South Australia  Australia  
952       115           South Jasper                    Queensland  Australia  
994      9386              New Harry             Western Australia  Australia

karena banyak informasi pada data yang sayang jika dibuang, maka gunakan metode imputation. Pada metode imputation, akan menggunakan nilai tertentu untuk mengganti missing value. Musabab kolom gender merupakan kolom kategorik, gunakan nilai yang dominan sebagai pengganti missing value tersebut. Gunakanlah method value_counts()untuk mengidentifikasi nilai yang dominan.

In [20]:
# identifikasi nilai dominan
customers_df.gender.value_counts()

Prefer not to say    725
Male                 143
Female               115
Name: gender, dtype: int64

In [21]:
# gunakan nilai dominan untuk mengganti missing value
customers_df.fillna(value="Prefer not to say", inplace=True)

In [22]:
# periksa kembali missing value
customers_df.isna().sum()

customer_id      0
customer_name    0
gender           0
age              0
home_address     0
zip_code         0
city             0
state            0
country          0
dtype: int64

In [23]:
# Menangani inaccurate value
# melihat data baris data yang mengandung inaccurate value tersebut
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
967          961     fulan 961  Prefer not to say  700   

                   home_address  zip_code        city            state  \
967  29 Farrell ParadeSuite 818      6528  New Joseph  South Australia   

       country  
967  Australia

Berdasarkan data tersebut, asumsikan bahwa inaccurate value tersebut terjadi karena human error sehingga kelebihan memasukkan nilai nol. Oleh karena itu, gantilah dengan nilai 70. Proses ini dilakukan dengan memanfaatkan method replace()

In [24]:
customers_df.age.replace(customers_df.age.max(), 70, inplace=True)

In [25]:
# periksa kembali baris data yang inacurrate
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
215          216     fulan 216  Prefer not to say  500   

               home_address  zip_code       city               state  \
215  038 Haley MewsApt. 810      3991  Bayertown  Northern Territory   

       country  
215  Australia

In [26]:
# perbaiki kembali data
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)

In [27]:
# periksa apakah masih ada inaccurate data?
customers_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    500.942058    49.874126  5000.693307
std     289.013599    17.644663  2886.084454
min       1.000000    20.000000     2.000000
25%     251.000000    34.000000  2398.000000
50%     501.000000    50.000000  5079.000000
75%     751.000000    65.000000  7454.000000
max    1000.000000    80.000000  9998.000000

Berdasarkan hasil tersebut dapat dilihat bahwa kolom age memiliki nilai maksimum yang cukup masuk akal. Selain itu, perhatikan, nilai mean dan standard deviation ikut berubah setelah kasus inaccurate value ditangani.

Selanjutnya Membersihkan Data orders_df.
Berdasarkan proses assessing data sebelumnya, diketahui bahwa terdapat kesalahan tipe data untuk kolom order_date & delivery_date.

Untuk mengatasi masalah ini, ganti tipe data pada kolom order_date & delivery_date menjadi datetime. Proses ini dapat dilakukan menggunakan function to_datetime() yang disediakan oleh library pandas.

In [28]:
# ubah tipe data menjadi datetime
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  orders_df[column] = pd.to_datetime(orders_df[column])

In [29]:
# periksa kembali tipe data dan jumlah data
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


Selanjutnya Membersihkan Data product_df.
Sesuai hasil assessing data sebelumnya, dapat diketahui bahwa terdapat 6 data yang terduplikasi pada product_df. Untuk mengatasi hal ini,perlu membuang data yang sama tersebut menggunakan method drop_duplicates()

In [30]:
product_df.drop_duplicates(inplace=True)

In [31]:
# pastikan kembali apakah masih ada duplikasi data tidak?
print("Jumlah duplikasi: ", product_df.duplicated().sum())

Jumlah duplikasi:  0


Selanjutnya membersihkan Data sales_df
Berdasarkan hasil penilaian data sebelumnya, diketahui bahwa terdapat 19 missing value pada kolom total_price. Untuk mengetahui proses penanganan missing value yang paling sesuai,perlu melihat terlebih dahulu baris data yang mengandung missing value tersebut.

In [32]:
sales_df[sales_df.total_price.isna()]

sales_id  order_id  product_id  price_per_unit  quantity  total_price
9            9         2        1196             105         1          NaN
121        121        27        1027              90         3          NaN
278        278        63         360              94         2          NaN
421        421        95        1091             115         1          NaN
489        489       108        1193             105         3          NaN
539        539       117         405             119         2          NaN
636        636       134         653              93         3          NaN
687        687       145        1138             102         1          NaN
854        854       177          64             104         1          NaN
1079      1079       222         908              94         3          NaN
1193      1193       248        1121             102         2          NaN
1313      1313       272         826             117         1          NaN
1548      1548       316         103             118         3          NaN
1688      1688       345         428             107         1          NaN
1775      1775       359         694             113         2          NaN
1902      1902       381        1218             105         3          NaN
2025      2025       408         611             112         3          NaN
2164      2164       436         583             100         3          NaN
2347      2347       476         696             113         2          NaN

Berdasarkan tampilan data tersebut, ditemukan bahwa nilai total_price merupakan hasil perkalian antara price_per_unit dan quantity. maka dapat gunakan pola ini untuk menangani missing value pada kolom total_price.

In [33]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

In [34]:
# periksa kembali jumlah missing value
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

Tujuan dari latihan ini untuk
membuat pertanyaan analisis atau bisnis yang ingin dicari jawabannya; dan
melakukan eksplorasi terhadap setiap data untuk mencari insight menarik guna menjawab pertanyaan bisnis tersebut.

masih menggunakan data dicoding collection yang sudah dilakukan pada latihan data wraling sebelumnya.


Menentukan Pertanyaan Bisnis
pertama, definisikan berbagai pertanyaan bisnis yang akan digunakan sebagai kompas dalam proses eksplorasi data, untuk menentukan sebuah pertanyaan, perlu memahami permasalahan bisnis yang dihadapi terlebih dahulu.


Sebagai perusahaan yang bergerak di bidang online fashion, DiCo perlu mengevaluasi performa penjualan (sales) dan memahami item fashion yang paling banyak dibeli. Selain itu, ia juga perlu lebih memahami pelanggannya (customer) sehingga dapat membuat sebuah strategi campaign yang lebih efisien.

Berdasarkan beberapa masalah tersebut, kita dapat mendefinisikan beberapa pertanyaan bisnis seperti berikut.

1. Bagaimana performa penjualan dan revenue perusahaan dalam beberapa bulan terakhir?
2. Produk apa yang paling banyak dan paling sedikit terjual?
3. Bagaimana demografi pelanggan yang dimiliki?
4. Kapan terakhir pelanggan melakukan transaksi?
5. Seberapa sering seorang pelanggan melakukan pembelian dalam beberapa bulan terakhir?
6. Berapa banyak uang yang dihabiskan pelanggan dalam beberapa bulan terakhir?

Melakukan Eksplorasi Data
lakukan eksplorasi data guna menjawab keenam pertanyaan tersebut.
Eksplorasi Data customers_df
dataset ini memuat berbagai informasi terkait customer, seperti customer_id, customer_name, gender, age, home_address, zip_code, city, state, dan country.

Sebagai permulaan, akan melihat rangkuman parameter statistik dari data customers_df menggunakan method describe().

In [83]:
print(customers_df.describe(include="all"))
print('---------------------------------------------------------------')
print(customers_df.sample(5))

        customer_id customer_name             gender          age  \
count   1001.000000          1001               1001  1001.000000   
unique          NaN          1000                  3          NaN   
top             NaN     fulan 943  Prefer not to say          NaN   
freq            NaN             2                743          NaN   
mean     500.942058           NaN                NaN    49.874126   
std      289.013599           NaN                NaN    17.644663   
min        1.000000           NaN                NaN    20.000000   
25%      251.000000           NaN                NaN    34.000000   
50%      501.000000           NaN                NaN    50.000000   
75%      751.000000           NaN                NaN    65.000000   
max     1000.000000           NaN                NaN    80.000000   

                      home_address     zip_code     city            state  \
count                         1001  1001.000000     1001             1001   
unique           

Berdasarkan hasil rangkuman parameter statistik diatas, diperoleh informasi jumlah pelanggan 1001 orang yang berumur kisaran 20 - 80 tahun dengan rata2 umur sebesar 49.87 tahun dengan standar deviasi sebesar 17.64 tahun.
Informasi ini memberikan gambaran bahwa segmentasi pelanggan yang dimiliki cukup luas mulai dari remaja hingga lanjut usia.

Selanjutnya cari tahu demografi pelanggan based on jenis kelamin (gender). dengan menggunakan method groupby() yang diikuti method agg()

In [36]:
customers_df.groupby(by="gender").agg({
    "customer_id": "nunique",
    "age": ["max", "min", "mean", "std"]
})

customer_id age                          
                      nunique max min       mean        std
gender                                                     
Female                    115  79  20  49.147826  16.646607
Male                      143  80  20  51.230769  18.462635
Prefer not to say         742  80  20  49.725437  17.644283

kode diatas menghasilkan sebuah pivot tabel yang berisi informasi jumlah pelanggan (nunique dari customer_id) serta parameter umur yang dikelompokkan berdasarkan jenis kelamin. berdasarkan pivot table diatas, diketahui bahwa pelanggan yang dimiliki didominasi oleh jenis kelamin prefer not to say. distribusi umur mereka ternyata cukup mirip berkisar 20 - 80 tahun.

selanjutnya, lihat persebaran jumlah pelanggan berdasarkan kota (city) dan negara bagian (state) menggunakan method groupby(). lalu urutkan nilainya menggunakan method sort_values() secara descending.

In [55]:
print(customers_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False))
print('------------------------------------------------------')
print(customers_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False))

city
East Aidan       3
East Sophia      3
New Ava          3
West Jackfort    2
Corkeryshire     2
                ..
Jordanside       1
Josephland       1
Josephmouth      1
Justinport       1
Zacville         1
Name: customer_id, Length: 961, dtype: int64
------------------------------------------------------
state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64


Berdasarkan hasil diatas, terlihat persebaran pelanggan cukup merata pada setiap kota dan negara bagian. Pelanggan paling banyak berada di kota East Aidan, East Sophia dan New Ava dengan masing2 jumlah 3 pelanggan. Selain itu pelanggan paling banyak berasal dari negara bagian South Australia.

Eksplorasi Data orders_df
mengandung berbagai informasi terkait sebuah order yang terdiri dari order_id, customer_id, order_date, dan delivery_date.
berdasarkan informasi tersebut, dapat dibuat kolom baru untuk memuat waktu pengiriman tiap order. untuk melakukannya, perlu menghitung selisih antar delivery_date dan order_date serta menyimpannya sebagai delivery_time.
lalu, gunakan method apply() untuk melakukan sebauh operasi terhadap setiap elemen dalam sebuah kolom dataframe atau series. Operasi yang akan dilakukan yaitu menghitung jumlah detik dari delivery_time dengan method total_seconds(). Nilai tersebut nantinya diubah kedalam satuan hari (dibagi 86400) dan diambil bilangan bulatnya saja.

In [84]:
orders_df.sample(5)

order_id  customer_id  payment order_date delivery_date  delivery_time
474       475          696    36229 2021-05-27    2021-05-27            0.0
626       627          680    19664 2021-04-03    2021-04-03            0.0
168       169          996    22089 2021-03-09    2021-03-09            0.0
255       256          313    15452 2021-08-09    2021-08-09            0.0
181       182          461    33720 2021-02-22    2021-02-22            0.0

In [86]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df["delivery_time"] = round(delivery_time/86400)

In [87]:
orders_df.describe(include="all", datetime_is_numeric=True)

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

                    delivery_date  delivery_time  
count                        1000         1000.0  
mean   2021-05-27 18:38:52.800000            0.0  
min           2021-01-01 00:00:00            0.0  
25%           2021-03-13 18:00:00            0.0  
50%           2021-05-27 12:00:00            0.0  
75%           2021-08-12 06:00:00            0.0  
max           2021-10-24 00:00:00            0.0  
std                           NaN            0.0

Berdasarkan hasil tersebut, dapat diketahui bahwa rata-rata waktu pengiriman sebesar 14 hari dengan nilai maksimum sebesar 27 hari dan nilai minimum sebesar 1 hari.

Eksplorasi Data orders_df dan customers_df

pada data orders_df terdapat kolom yang berisi informasi terkait customer id pelanggan yang pernah melakukan order. Informasi ini bisa digunakan untuk mengidentifikasi pelanggan yang belum pernah melakukan order. Untuk melakukan hal ini, bisa membuat sebuah kolom baru bernama “status” pada data customers_df. Kolom tersebut memiliki nilai “Active” untuk pelanggan yang pernah melakukan order setidaknya sekali dan sebaliknya bernilai “Non Active” untuk pelanggan yang belum pernah melakukan order sama sekali.

In [64]:
customer_id_in_orders_df = orders_df.customer_id.tolist()
customers_df["status"] = customers_df["customer_id"].apply(lambda x: "Active" if x in customer_id_in_orders_df else "Non Active")
customers_df.sample(5)

customer_id customer_name             gender  age  \
741          739     fulan 739             Female   57   
855          852     fulan 852  Prefer not to say   67   
373          373     fulan 373  Prefer not to say   44   
32            33      fulan 33               Male   44   
628          627     fulan 627  Prefer not to say   51   

                       home_address  zip_code               city  \
741         09 Jesse ParadeApt. 250      1335  East Matthewville   
855          079 Chase MallApt. 902      8460       O'connertown   
373  7319 O'neill JunctionSuite 661      8769       North Sienna   
32        639 Farrell HillSuite 093      4116         Chaseshire   
628        66 Isla CrescentApt. 573      7404   Port Christopher   

                            state    country      status  
741               New South Wales  Australia      Active  
855                      Tasmania  Australia      Active  
373  Australian Capital Territory  Australia  Non Active  
32                South Australia  Australia  Non Active  
628               South Australia  Australia      Active

untuk memperoleh inforamsi jumlah pelanggan yang berstatus aktif dan tidak aktif bisa gunakan pivot table.

In [65]:
customers_df.groupby(by="status").customer_id.count()

status
Active        617
Non Active    384
Name: customer_id, dtype: int64

dari hasil diatas, terdapat cukup banyak pelanggan yang belum pernah melakukan transaksi sebelumnya. hampir 30% pelanggan belum pernah melakukan order sebelumnya. untuk memperoleh informasi lebih banyak terkait 2 data tsb, perlu gabungkan keduanya melalui proses merge atau join.

In [66]:
orders_customers_df = pd.merge(
    left=orders_df,
    right=customers_df,
    how="left",
    left_on="customer_id",
    right_on="customer_id"
)

orders_customers_df.head()

order_id  customer_id  payment order_date delivery_date  delivery_time  \
0         1           64    30811 2021-08-30    2021-08-30            0.0   
1         2          473    50490 2021-02-03    2021-02-03            0.0   
2         3          774    46763 2021-10-08    2021-10-08            0.0   
3         4          433    39782 2021-05-06    2021-05-06            0.0   
4         5          441    14719 2021-03-23    2021-03-23            0.0   

  customer_name             gender  age                   home_address  \
0      fulan 64  Prefer not to say   75      4927 Alice MeadowApt. 960   
1     fulan 473               Male   61  531 Schmitt BoulevardApt. 010   
2     fulan 774  Prefer not to say   34       2096 Wilson MewsApt. 714   
3     fulan 433  Prefer not to say   46        5777 Mayer PassApt. 881   
4     fulan 441  Prefer not to say   53   33 Richards JunctionApt. 478   

   zip_code                city              state    country  status  
0      7787      Sanfordborough    South Australia  Australia  Active  
1      1744             Annaton    South Australia  Australia  Active  
2      8590  West Jonathanshire           Tasmania  Australia  Active  
3      9728    West Michaelport           Tasmania  Australia  Active  
4      7650     South Rileyview  Western Australia  Australia  Active

pada kode diatas, proses merge dengan metode "left" dilakukan untuk mengambil semua nilai dari tabel kiri serta nilai yang bersesuaian dengan tabel kanan.
terdapat banyak sekali informasi yang bisa di eksplore.

In [67]:
# jumlah order berdasarkan kota
orders_customers_df.groupby(by="city").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

city  order_id
0        Jordanside         6
1           New Ava         6
2         Lake Rose         5
3        O'keefeton         5
4          East Max         5
5  Port Hannahburgh         5
6          Rubyfort         5
7          West Kai         5
8     Andrewborough         4
9     Port Aaronton         4

dari data diatas, diketahui kota jordanside dan new ava merupakan kota yang memiliki jumlah order terbanyak.

In [68]:
# jumlah order berdasarkan state
orders_customers_df.groupby(by="state").order_id.nunique().sort_values(ascending=False)

state
South Australia                 148
Queensland                      139
Western Australia               130
New South Wales                 129
Australian Capital Territory    118
Victoria                        118
Tasmania                        112
Northern Territory              106
Name: order_id, dtype: int64

diketahui dari data diatas bahwa south australia merupakan state yang paling banyak melakukan order.

In [69]:
# jumlah order berdasarkan gender
orders_customers_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False)

gender
Prefer not to say    725
Female               139
Male                 136
Name: order_id, dtype: int64

berdasarkan hasil tersebut, diketahui kebanyakan order dilakukan oleh pelanggan yang berjenis kelamin prefer not to say. sejalan dengan jumlah pelanggan yang kebanyakan dari kelompok gender tsb.

In [70]:
# jumlah order berdasarkan kelompok usia
# definisikan kolom baru bernama age_group
# kelompokan pelanggan jadi 3 kelompok (youth, adults, seniors)
orders_customers_df["age_group"] = orders_customers_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))
orders_customers_df.groupby(by="age_group").order_id.nunique().sort_values(ascending=False)

age_group
Adults     681
Seniors    226
Youth       93
Name: order_id, dtype: int64

berdasarkan data diatas, diketahui bahwa pelanggan paling banyak melakukan order adalah kelompok usia adults.

Eksplorasi data product_df dan sales_df
pertama, lihat rangkuman parameter statistik dari kedua nya dengan menggunakan method describe()

In [72]:
print(product_df.describe(include="all"))
print('-----------------------------------------------------------------------------------------------------------------------------------')
print(sales_df.describe(include="all"))

         product_id product_type product_name  size colour        price  \
count   1260.000000         1260         1260  1260   1260  1260.000000   
unique          NaN            3           35     8      7          NaN   
top             NaN        Shirt        Denim    XS    red          NaN   
freq            NaN          420           70   252    180          NaN   
mean     629.500000          NaN          NaN   NaN    NaN   105.805556   
std      363.874979          NaN          NaN   NaN    NaN     9.704423   
min        0.000000          NaN          NaN   NaN    NaN    90.000000   
25%      314.750000          NaN          NaN   NaN    NaN    95.750000   
50%      629.500000          NaN          NaN   NaN    NaN   108.500000   
75%      944.250000          NaN          NaN   NaN    NaN   114.000000   
max     1259.000000          NaN          NaN   NaN    NaN   119.000000   

           quantity                                   description  
count   1260.000000            

Berdasarkan hasil tersebut, dapat dilihat bahwa harga barang yang dijual berkisar antara 90 hingga 119 dolar. Selain itu, pada setiap transaksi pelanggan paling banyak membeli tiga buah item dalam satu jenis produk dengan total harga sebesar 357 dolar.

In [73]:
# ketahui harga produk termahal dan termurah
product_df.sort_values(by="price", ascending=False)

product_id product_type product_name size  colour  price  quantity  \
704          698       Jacket        Parka    L  violet    119        53   
671          665       Jacket        Parka   XS     red    119        65   
698          692       Jacket        Parka    M  indigo    119        66   
699          693       Jacket        Parka    L  indigo    119        44   
700          694       Jacket        Parka   XL  indigo    119        78   
...          ...          ...          ...  ...     ...    ...       ...   
1225        1219     Trousers      Pleated   XL  indigo     90        45   
1226        1220     Trousers      Pleated   XS  violet     90        58   
1227        1221     Trousers      Pleated    S  violet     90        50   
1228        1222     Trousers      Pleated    M  violet     90        45   
599          593       Jacket       Bomber    L  violet     90        68   

                                        description  
704        A violet coloured, L sized, Parka Jacket  
671          A red coloured, XS sized, Parka Jacket  
698        A indigo coloured, M sized, Parka Jacket  
699        A indigo coloured, L sized, Parka Jacket  
700       A indigo coloured, XL sized, Parka Jacket  
...                                             ...  
1225  A indigo coloured, XL sized, Pleated Trousers  
1226  A violet coloured, XS sized, Pleated Trousers  
1227   A violet coloured, S sized, Pleated Trousers  
1228   A violet coloured, M sized, Pleated Trousers  
599       A violet coloured, L sized, Bomber Jacket  

[1260 rows x 8 columns]

gunakan pivot table untuk mencari informasi terkait produk berdasarkan tipe dan nama produknya.

In [75]:
print(product_df.groupby(by="product_type").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
}))

print('-----------------------------------------------------------------------')

print(
 product_df.groupby(by="product_name").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})
)

             product_id quantity price     
                nunique      sum   min  max
product_type                               
Jacket              420    25387    90  119
Shirt               420    25263    92  119
Trousers            420    25139    90  119
-----------------------------------------------------------------------
                  product_id quantity price     
                     nunique      sum   min  max
product_name                                    
Bomber                    35     2083    90   90
Camp Collared             35     2071   112  112
Cardigan                  35     2032   118  118
Cargo Pants               35     2146   106  106
Casual Slim Fit           35     2086   113  113
Chambray                  35     2020   105  105
Chinos                    35     2101   100  100
Coach                     35     2158   115  115
Cords                     35     2260   113  113
Cropped                   35     2085    99   99
Cuban Collar              3

In [76]:
# merge tabel product_df dan sales_df utk mengetahui produk terlaris
sales_product_df = pd.merge(
    left=sales_df,
    right=product_df,
    how="left",
    left_on="product_id",
    right_on="product_id"
)

sales_product_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  price  quantity_y  \
0        Shirt      Chambray    L  orange    105          44   
1       Jacket        Puffer    S  indigo    110          62   
2        Shirt  Oxford Cloth    M     red    114          54   
3     Trousers          Wool    M    blue    111          52   
4       Jacket         Parka    S  indigo    119          53   

                                   description  
0   A orange coloured, L sized, Chambray Shirt  
1    A indigo coloured, S sized, Puffer Jacket  
2  A red coloured, M sized, Oxford Cloth Shirt  
3      A blue coloured, M sized, Wool Trousers  
4     A indigo coloured, S sized, Parka Jacket

hasil dari proses merge memiliki perbedaan antara nilai price_per_unit dan price. hal ini bisa terjadi karena adanya potongan harga, biaya operasional, dan biaya lainnya.

selanjutnya lihat informasi penjualan berdasarkan tipe.

In [77]:
sales_product_df.groupby(by="product_type").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
})

sales_id  quantity_x  total_price
product_type                                   
Jacket            1676        3343       357026
Shirt             1641        3259       333600
Trousers          1683        3360       341174

perhatikan, trousers adalah tipe produk yang paling laris terjual. tapi jika diliaht berdasarkan revenue yang diterima, jacket merupakan tipe produk yang paling banyak menyumbang revenue ke perusahaan.

In [78]:
# buat pivot table untuk informasi penjualan based on nama produk

sales_product_df.groupby(by="product_name").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
}).sort_values(by="total_price", ascending=False)

sales_id  quantity_x  total_price
product_name                                        
Denim                   273         527        52399
Casual Slim Fit         154         306        36414
Trench Coat             146         299        35581
Shearling               150         302        35334
Puffer                  140         298        35164
Flannel                 141         281        33158
Cropped                 135         284        32660
Pleated                 147         308        32340
Joggers                 164         334        31062
Chambray                141         290        30740
Parka                   134         269        30397
Bomber                  150         297        29700
Chinos                  135         260        29380
Windbreaker             143         287        29274
Cardigan                134         260        29120
Peacoat                 132         266        28994
Tracksuit Bottoms       130         257        28784
Slim-Fit                154         288        27936
Oxford Cloth            146         289        27744
Drawstring              147         288        27648
High-Waisted            140         278        27522
Camp Collared           139         279        27342
Cargo Pants             134         267        27234
Relaxed Leg             146         301        27090
Cuban Collar            122         254        26670
Henley                  146         281        26414
Wool                    125         249        26394
Leather                 136         263        25774
Mandarin Collar         124         236        25724
Dress                   127         243        25272
Pullover                143         276        25116
Polo                    131         264        25080
Coach                   134         265        24645
Linen                   131         270        24570
Cords                   126         246        23124

dari hasil data diatas, diketahui bahwa produk denim merupakan produk yang paling laris dan paling banyak menyumbang revenue ke perusahaan.

Eksplorasi Data  all_df
untuk melihat pola pembelian berdasarkan demografi pelanggan. buat sebuah DataFrame baru bernama all_df untuk menampung semua informasi dari keempat tabel yang ada.

In [79]:
all_df = pd.merge(
    left=sales_product_df,
    right=orders_customers_df,
    how="left",
    left_on="order_id",
    right_on="order_id"
)

all_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  ...  customer_name  \
0        Shirt      Chambray    L  orange  ...       fulan 64   
1       Jacket        Puffer    S  indigo  ...       fulan 64   
2        Shirt  Oxford Cloth    M     red  ...       fulan 64   
3     Trousers          Wool    M    blue  ...       fulan 64   
4       Jacket         Parka    S  indigo  ...       fulan 64   

              gender age               home_address  zip_code            city  \
0  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
1  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
2  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
3  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
4  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   

             state    country  status age_group  
0  South Australia  Australia  Active   Seniors  
1  South Australia  Australia  Active   Seniors  
2  South Australia  Australia  Active   Seniors  
3  South Australia  Australia  Active   Seniors  
4  South Australia  Australia  Active   Seniors  

[5 rows x 28 columns]

selanjutnya coba lihat preferensi pembelian berdasarkan state pelanggan dan tipe product

In [80]:
all_df.groupby(by=["state", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
state                        product_type                         
Australian Capital Territory Jacket               406        43204
                             Shirt                396        40448
                             Trousers             454        46790
New South Wales              Jacket               451        47998
                             Shirt                431        43980
                             Trousers             392        39766
Northern Territory           Jacket               365        38991
                             Shirt                336        33865
                             Trousers             384        38998
Queensland                   Jacket               499        53511
                             Shirt                417        42506
                             Trousers             453        46045
South Australia              Jacket               461        49090
                             Shirt                509        52685
                             Trousers             455        46041
Tasmania                     Jacket               412        44370
                             Shirt                387        39668
                             Trousers             379        38057
Victoria                     Jacket               359        38203
                             Shirt                390        40206
                             Trousers             414        41948
Western Australia            Jacket               390        41659
                             Shirt                393        40242
                             Trousers             429        43529

dari data diatas memberikan gambaran terkait tipe produk yang disukai pengguna berdasarkan lokasi negara bagiannya. Jika diperhatikan, tipe produk Jacket (garis merah) paling banyak terjual pada negara bagian Queensland, South Australia, dan New South Wales.

selanjutnya, ketahui selera tipe produk pelanggan berdasarkan gender dan kelompok usia.

In [81]:
print(all_df.groupby(by=["gender", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
}))

print('-------------------------------------------------------------------------')


print(all_df.groupby(by=["age_group", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
}))

                                quantity_x  total_price
gender            product_type                         
Female            Jacket               481        50963
                  Shirt                445        45146
                  Trousers             481        49295
Male              Jacket               480        50978
                  Shirt                410        41939
                  Trousers             453        46071
Prefer not to say Jacket              2382       255085
                  Shirt               2404       246515
                  Trousers            2426       245808
-------------------------------------------------------------------------
                        quantity_x  total_price
age_group product_type                         
Adults    Jacket              2292       245055
          Shirt               2225       227781
          Trousers            2272       231271
Seniors   Jacket               777        82959
          Shirt       

dari hasil pivot table diatas dapat memberikan gambaran terkait selera tipe produk pelanggan berdasarkan gender dan kelompok usia.